### ISSUE TITLE
Ambient declarations not working due to import/export

### ISSUE THREAD

**Issue by carmanchris31** [danger/danger-js#955]
The `import` and `export` statements in `danger.d.ts` are preventing the danger DSL from being available in the global scope:

![danger-ambient-declarations](https://user-images.githubusercontent.com/9935648/68442571-4bf7d500-0186-11ea-908f-f8e30ace69a8.gif)

This is a known issue in Typescript `*.d.ts` files (ambient declarations):
https://github.com/microsoft/TypeScript/issues/28097

A potential solution per [this comment](https://github.com/microsoft/TypeScript/issues/28097#issuecomment-443509326) is to wrap global declarations in a `declare global { }` block.

Where should one look in the code in order to attempt addressing this?

**Comment by orta** [comment]
I specifically choose to not have them available in the global scope, can you explain your reasoning for why you'd want that?

**Comment by carmanchris31** [comment]
Since they are, in fact, provided globally, similar to methods like `describe`/`expect` from jest.

In the [documentation for plugins](https://danger.systems/js/usage/extending-danger.html#writing-your-plugin), it's explicitly stated that danger should not be imported (emphasis added):
> Note, that you should not add a import { danger, et, cetera } from "danger" in your plugin. The dirty truth is that Danger completely ignores that at runtime, and just **puts all of it’s exports inside the global context**. You’ll get compiler errors, or confusing shadow variable issues. **So assume that it’s there**, and working fine.

The emphasized portions in particular aligns with the expectation that, in a TypeScript environment, I should have access to the ambient declarations. Since they aren't available, I'm left with TypeScript errors. If I import from `danger`, I get errors from the `danger` cli. It creates a painful dilemma.

**Comment by orta** [comment]
Makes sense - how we deal with this is:

[in tests](https://github.com/orta/danger-plugin-lighthouse/blob/master/src/index.test.ts#L3-L19)
```ts
declare const global: any;

describe("lighthouse()", () => {
  beforeEach(() => {
    global.warn = jest.fn();
    global.message = jest.fn();
    global.fail = jest.fn();
    global.markdown = jest.fn();
  });

  afterEach(() => {
    global.warn = undefined;
    global.message = undefined;
    global.fail = undefined;
    global.markdown = undefined;
  });

```

and [in app-code](https://github.com/orta/danger-plugin-lighthouse/blob/master/src/index.ts#L1-L8)

```ts
// Provides dev-time type structures for  `danger` - doesn't affect runtime.
import { DangerDSLType } from "../node_modules/danger/distribution/dsl/DangerDSL"
declare var danger: DangerDSLType
export declare function message(message: string): void
export declare function warn(message: string): void
export declare function fail(message: string): void
export declare function markdown(message: string): void
```

Which gets set up via the yeoman generator for plugins

**Comment by kilahm** [comment]
I may be digging up an old thread so feel free to lock and not respond.

@orta - could you explain your reasoning behind forcing consumers to declare the globally available functions provided by Danger?

Having an import from so deep in the package, with an explicit reference to `node_modules` smells of reaching into the internals, which could (and should) change without creating BC breaks.

I'd be happy to work on a PR to make the declaration of the functions and constants global, allowing this `tsconfig.json` to work:

```json
{
  "compilerOptions": {
    "target": "es6",
    "module": "es2015",
    "moduleResolution": "Node",
    "types": ["danger", "node"]
  },
  "include": ["./*"]
}
```

**Comment by orta** [comment]
Yeah, I'm probably not changing my mind on this one. Only plugin authors have to do the hack, doesn't pollute global scope for users and it fits how people would expect it to work in a dangerfile.

**Comment by fbartho** [comment]
We could provide a file that plugin authors could directly reference that would hook up that global scope?

Danger-plugin-interface.d.ts (not sure if that would work?)

**Comment by carmanchris31** [comment]
Danger's very own DangerFile contains a substantial amount of declaration boilerplate as a workaround for this issue:
https://github.com/danger/danger-js/blob/master/dangerfile.ts#L1-L17

That's not what I'd expect to have to do every time I write a dangerfile that does something other than use existing plugins.

Perhaps an acceptable way of accomplishing this would be with two steps:
1. Add the action methods (`warn`, `fail`, `comment`, `schedule`, etc.) to Danger's DSL;
  these can be namespaced if desired, e.g. `danger.actions.fail()`
2. Export `DangerDSLType` from the main package
  Currently we can do `import {DangerDSLType} from 'danger/distribution/dsl/DangerDSL';` but as @kilahm noted that's not ideal
3. (long-term) Deprecate the use of the globally-injected actions in favor of the DSL-scoped versions.

The result of this would be that all consumers of danger could simply do this:
```js
import {DangerDSLType} from 'danger';

declare const danger: DangerDSLType;
```
and have full access to the fully-typed Danger API.

**Comment by orta** [comment]
You should never need to include those helpers in user-land Dangerfile

That said, I think re-exporting the types from an index file is a cool idea 👍🏻

**Comment by airtonix** [comment]
> You should never need to include those helpers in user-land Dangerfile
> 
> That said, I think re-exporting the types from an index file is a cool idea 👍🏻

wat? 

![image](https://user-images.githubusercontent.com/61225/156304073-b8436dfe-88cf-430b-9f30-f35ba9d0fd17.png)
